File: solutions/opus-1/opus/ruby/dijkstra.rb
data = STDIN.read.split.map(&:to_i)
n, m = data[0], data[1]
adj = Array.new(n) { [] }
i = 2
m.times do
  u, v, w = data[i], data[i + 1], data[i + 2]
  i += 3
  adj[u] << [v, w]
  adj[v] << [u, w]
end
s, t = data[i], data[i + 1]

if s == t
  puts 0
  exit
end

dist = Array.new(n)
dist[s] = 0
heap = [[0, s]]

push = lambda do |item|
  heap << item
  k = heap.size - 1
  while k > 0
    p = (k - 1) >> 1
    break if heap[p][0] <= heap[k][0]
    heap[p], heap[k] = heap[k], heap[p]
    k = p
  end
end

pop = lambda do
  top = heap[0]
  last = heap.pop
  unless heap.empty?
    heap[0] = last
    k = 0
    size = heap.size
    loop do
      l = 2 * k + 1
      break if l >= size
      r = l + 1
      c = (r < size && heap[r][0] < heap[l][0]) ? r : l
      break if heap[k][0] <= heap[c][0]
      heap[k], heap[c] = heap[c], heap[k]
      k = c
    end
  end
  top
end

ans = -1
until heap.empty?
  d, u = pop.call
  next if d > dist[u]
  if u == t
    ans = d
    break
  end
  adj[u].each do |v, w|
    nd = d + w
    if dist[v].nil? || nd < dist[v]
      dist[v] = nd
      push.call([nd, v])
    end
  end
end

puts ans

File: solutions/opus-1/opus/perl/dijkstra.pl
use strict;
use warnings;

local $/;
my $in = <STDIN>;
my @t = split ' ', $in;
my $p = 0;
my ($n, $m) = @t[$p, $p + 1]; $p += 2;
my @adj = map { [] } 0 .. $n - 1;
for (1 .. $m) {
    my ($u, $v, $w) = @t[$p .. $p + 2]; $p += 3;
    push @{ $adj[$u] }, $v, $w;
    push @{ $adj[$v] }, $u, $w;
}
my ($s, $e) = @t[$p, $p + 1];

if ($s == $e) { print "0\n"; exit; }

my @dist;
$dist[$s] = 0;
my @hd = (0);
my @hn = ($s);

while (@hd) {
    my $d = $hd[0];
    my $u = $hn[0];
    my $ld = pop @hd;
    my $ln = pop @hn;
    if (@hd) {
        my $i = 0;
        my $sz = @hd;
        while (1) {
            my $c = 2 * $i + 1;
            last if $c >= $sz;
            $c++ if $c + 1 < $sz && $hd[$c + 1] < $hd[$c];
            last if $hd[$c] >= $ld;
            $hd[$i] = $hd[$c];
            $hn[$i] = $hn[$c];
            $i = $c;
        }
        $hd[$i] = $ld;
        $hn[$i] = $ln;
    }
    next if $d > $dist[$u];
    if ($u == $e) { print "$d\n"; exit; }
    my $a = $adj[$u];
    for (my $k = 0; $k < @$a; $k += 2) {
        my $v  = $a->[$k];
        my $nd = $d + $a->[$k + 1];
        if (!defined $dist[$v] || $nd < $dist[$v]) {
            $dist[$v] = $nd;
            my $i = @hd;
            while ($i > 0) {
                my $pi = ($i - 1) >> 1;
                last if $hd[$pi] <= $nd;
                $hd[$i] = $hd[$pi];
                $hn[$i] = $hn[$pi];
                $i = $pi;
            }
            $hd[$i] = $nd;
            $hn[$i] = $v;
        }
    }
}
print "-1\n";
